<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Detecting signatures of selection using KLFDAPC • KLFDAPC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Detecting signatures of selection using KLFDAPC">
<meta property="og:description" content="KLFDAPC">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-158177382-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-158177382-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">KLFDAPC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../reference/index.html">Manual</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinghuq/KLFDAPC/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Detecting signatures of selection using KLFDAPC</h1>
                        <h4 class="author">Xinghu Qin</h4>
            
            <h4 class="date">27/10/2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinghuq/KLFDAPC/blob/master/vignettes/Genome_scan_KLFDAPC.Rmd"><code>vignettes/Genome_scan_KLFDAPC.Rmd</code></a></small>
      <div class="hidden name"><code>Genome_scan_KLFDAPC.Rmd</code></div>

    </div>

    
    
<div id="klfdapc-for-detection-of-natural-selection" class="section level2">
<h2 class="hasAnchor">
<a href="#klfdapc-for-detection-of-natural-selection" class="anchor"></a>KLFDAPC for detection of natural selection</h2>
<p>KLFDAPC has multiple advantages over the linear inference and linear discriminant analysis. The population structure inferred using KLFDAPC rectifies the limitation of linear population structure. Thereby, KLFDAPC could be helpful for correcting the genetic gradients (population stratification) in many aspects of population genetic studies, such as correction for geographic genetic structure in genome scan for detecting loci involved in adaptation, and in GWAS for identifying the signatures related to diseases or complex traits. We demonstrate an example using KLFDAPC to detect signatures related to local adaptation. In what follows, we illustrate the step-by-step implementation of genome scan using KLFDAPC.</p>
<p>In this vignette, we use a common implementation similar to PCA-based genome scan that reported by Patterson, N.,et al. 2006 and Luu, K., et al. 2017. However, in this implementation, the loci are weighted by correlation coefficients. We recommended users using more sophisticated implementation with high performance, such as DeepGenomeScan (Qin,X. et al. 2020).</p>
<p>Install packages and dependences</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"SNPRelate"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"SNPRelate"</span><span class="op">)</span>

<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"KLFDAPC"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/devtools/man/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"xinghuq/KLFDAPC"</span><span class="op">)</span>

<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"DA"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/devtools/man/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"xinghuq/DA"</span><span class="op">)</span></pre></div>
<p>Our package uses the high performance parallel computation processed with SNPRelate. It is written by C++, and the data is stored in an efficient format, snpGDS format.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/zhengxwen/SNPRelate">SNPRelate</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinghuq/KLFDAPC">KLFDAPC</a></span><span class="op">)</span></pre></div>
<div id="preparing-data" class="section level3">
<h3 class="hasAnchor">
<a href="#preparing-data" class="anchor"></a>Preparing data</h3>
<p>In this vignette, We use the human HapMap data (<a href="https://www.genome.gov/10001688/international-hapmap-project" class="uri">https://www.genome.gov/10001688/international-hapmap-project</a>) to show how to detect loci involved in adaptation using KLFDAPC.</p>
<p>Data preparation, we use the data that have already in GDS format stored in the library.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="co"># Open the GDS file</span>
<span class="va">genofile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html">snpgdsOpen</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsExampleFileName.html">snpgdsExampleFileName</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co"># Get sample id</span>
<span class="va">sample.id</span> <span class="op">&lt;-</span> <span class="fu">read.gdsn</span><span class="op">(</span><span class="fu">index.gdsn</span><span class="op">(</span><span class="va">genofile</span>, <span class="st">"sample.id"</span><span class="op">)</span><span class="op">)</span>
<span class="co"># Get population information</span>

<span class="va">pop_code</span> <span class="op">&lt;-</span> <span class="fu">read.gdsn</span><span class="op">(</span><span class="fu">index.gdsn</span><span class="op">(</span><span class="va">genofile</span>, <span class="st">"sample.annot/pop.group"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># assume the order of sample IDs is as the same as population codes</span>

<span class="va">pop_code</span> <span class="op">&lt;-</span> <span class="fu">read.gdsn</span><span class="op">(</span><span class="fu">index.gdsn</span><span class="op">(</span><span class="va">genofile</span>, path<span class="op">=</span><span class="st">"sample.annot/pop.group"</span><span class="op">)</span><span class="op">)</span>
<span class="va">pop_code</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">pop_code</span>,levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">pop_code</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">pop_code</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span>

<span class="co"># fliter LD thresholds for sensitivity analysis</span>
<span class="va">snpset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsLDpruning.html">snpgdsLDpruning</a></span><span class="op">(</span><span class="va">genofile</span>, ld.threshold<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span>

<span class="co"># Get all selected snp id</span>
<span class="va">snpset.id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unname.html">unname</a></span><span class="op">(</span><span class="va">snpset</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">snpset.id</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsClose.html">snpgdsClose</a></span><span class="op">(</span><span class="va">genofile</span><span class="op">)</span>
<span class="co"># pop_code</span>
<span class="co"># YRI CEU HCB JPT </span>
<span class="co">#  93  92  47  47 </span>
<span class="co"># SNP pruning based on LD:</span>
<span class="co"># Excluding 365 SNPs on non-autosomes</span>
<span class="co"># Excluding 1 SNP (monomorphic: TRUE, MAF: NaN, missing rate: NaN)</span>
<span class="co"># Working space: 279 samples, 8,722 SNPs</span>
<span class="co">#     using 1 (CPU) core</span>
<span class="co">#     sliding window: 500,000 basepairs, Inf SNPs</span>
<span class="co">#     |LD| threshold: 0.2</span>
<span class="co">#     method: composite</span>
<span class="co"># Chromosome 1: 76.12%, 545/716</span>
<span class="co"># Chromosome 2: 72.78%, 540/742</span>
<span class="co"># Chromosome 3: 74.71%, 455/609</span>
<span class="co"># Chromosome 4: 73.49%, 413/562</span>
<span class="co"># Chromosome 5: 76.86%, 435/566</span>
<span class="co"># Chromosome 6: 75.75%, 428/565</span>
<span class="co"># Chromosome 7: 75.42%, 356/472</span>
<span class="co"># Chromosome 8: 71.11%, 347/488</span>
<span class="co"># Chromosome 9: 77.88%, 324/416</span>
<span class="co"># Chromosome 10: 74.12%, 358/483</span>
<span class="co"># Chromosome 11: 77.85%, 348/447</span>
<span class="co"># Chromosome 12: 76.81%, 328/427</span>
<span class="co"># Chromosome 13: 76.16%, 262/344</span>
<span class="co"># Chromosome 14: 76.60%, 216/282</span>
<span class="co"># Chromosome 15: 76.34%, 200/262</span>
<span class="co"># Chromosome 16: 72.66%, 202/278</span>
<span class="co"># Chromosome 17: 73.91%, 153/207</span>
<span class="co"># Chromosome 18: 73.68%, 196/266</span>
<span class="co"># Chromosome 19: 85.00%, 102/120</span>
<span class="co"># Chromosome 20: 71.62%, 164/229</span>
<span class="co"># Chromosome 21: 76.98%, 97/126</span>
<span class="co"># Chromosome 22: 75.86%, 88/116</span>
<span class="co"># 6,557 markers are selected in total.</span>
<span class="co"># [1]  1  2  4  5  7 10</span></pre></div>
</div>
<div id="doing-klfdapc-on-genotype-data-with-a-gaussian-kernel" class="section level3">
<h3 class="hasAnchor">
<a href="#doing-klfdapc-on-genotype-data-with-a-gaussian-kernel" class="anchor"></a>Doing KLFDAPC on genotype data with a Gaussian kernel</h3>
<p>We use a Gaussian kernel with sigma =5 to infer the population structure. After we obtain the genetic structure, we then use it to do genome scan.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="kw">showfile.gds</span>(<span class="dt">closeall=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb4-2" title="2"></a>
<a class="sourceLine" id="cb4-3" title="3">klfdapcfile &lt;-<span class="st"> </span><span class="kw">KLFDAPC</span>(<span class="kw">snpgdsExampleFileName</span>(), <span class="dt">y=</span>pop_code, <span class="dt">snp.id=</span>snpset.id,<span class="dt">n.pc =</span> <span class="dv">3</span>,<span class="dt">kernel=</span>kernlab<span class="op">::</span><span class="kw">rbfdot</span>(<span class="dt">sigma =</span> <span class="dv">5</span>), <span class="dt">r=</span><span class="dv">5</span>, <span class="dt">num.thread=</span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb4-4" title="4"><span class="co"># Warning: package 'klaR' was built under R version 3.6.3</span></a>
<a class="sourceLine" id="cb4-5" title="5"></a>
<a class="sourceLine" id="cb4-6" title="6"><span class="co"># Make a data.frame</span></a>
<a class="sourceLine" id="cb4-7" title="7">pop_str &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">sample.id =</span> klfdapcfile<span class="op">$</span>PCA<span class="op">$</span>sample.id,</a>
<a class="sourceLine" id="cb4-8" title="8">    <span class="dt">pop =</span> <span class="kw">factor</span>(pop_code)[<span class="kw">match</span>(klfdapcfile<span class="op">$</span>PCA<span class="op">$</span>sample.id, sample.id)],</a>
<a class="sourceLine" id="cb4-9" title="9">    <span class="dt">RD1 =</span> klfdapcfile<span class="op">$</span>KLFDAPC<span class="op">$</span>Z[,<span class="dv">1</span>],    <span class="co"># the first reduced feature</span></a>
<a class="sourceLine" id="cb4-10" title="10">    <span class="dt">RD2 =</span>  klfdapcfile<span class="op">$</span>KLFDAPC<span class="op">$</span>Z[,<span class="dv">2</span>],    <span class="co"># the second reduced feature</span></a>
<a class="sourceLine" id="cb4-11" title="11">    <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb4-12" title="12"></a>
<a class="sourceLine" id="cb4-13" title="13"><span class="kw">head</span>(pop_str)</a>
<a class="sourceLine" id="cb4-14" title="14"></a>
<a class="sourceLine" id="cb4-15" title="15"><span class="kw">rownames</span>(pop_str)=pop_str<span class="op">$</span>sample.id</a>
<a class="sourceLine" id="cb4-16" title="16"><span class="co"># Principal Component Analysis (PCA) on genotypes:</span></a>
<a class="sourceLine" id="cb4-17" title="17"><span class="co"># Excluding 2,531 SNPs (non-autosomes or non-selection)</span></a>
<a class="sourceLine" id="cb4-18" title="18"><span class="co"># Excluding 0 SNP (monomorphic: TRUE, MAF: NaN, missing rate: NaN)</span></a>
<a class="sourceLine" id="cb4-19" title="19"><span class="co"># Working space: 279 samples, 6,557 SNPs</span></a>
<a class="sourceLine" id="cb4-20" title="20"><span class="co">#     using 2 (CPU) cores</span></a>
<a class="sourceLine" id="cb4-21" title="21"><span class="co"># PCA:    the sum of all selected genotypes (0,1,2) = 1828045</span></a>
<a class="sourceLine" id="cb4-22" title="22"><span class="co"># CPU capabilities: Double-Precision SSE2</span></a>
<a class="sourceLine" id="cb4-23" title="23"><span class="co"># Thu Oct 29 00:40:50 2020    (internal increment: 1276)</span></a>
<a class="sourceLine" id="cb4-24" title="24"><span class="co"># </span></a>
<a class="sourceLine" id="cb4-25" title="25">[..................................................]  <span class="dv">0</span>%, ETC<span class="op">:</span><span class="st"> </span><span class="op">---</span><span class="st">        </span></a>
<a class="sourceLine" id="cb4-26" title="26">[<span class="op">==</span><span class="er">================================================</span>] <span class="dv">100</span>%, completed, 0s</a>
<a class="sourceLine" id="cb4-27" title="27"><span class="co"># Thu Oct 29 00:40:50 2020    Begin (eigenvalues and eigenvectors)</span></a>
<a class="sourceLine" id="cb4-28" title="28"><span class="co"># Thu Oct 29 00:40:50 2020    Done.</span></a>
<a class="sourceLine" id="cb4-29" title="29"><span class="co"># [1] "Doing KLFDAPC"</span></a>
<a class="sourceLine" id="cb4-30" title="30"><span class="co">#   sample.id pop       RD1       RD2</span></a>
<a class="sourceLine" id="cb4-31" title="31"><span class="co"># 1   NA19152 YRI -53.71177 -19.85460</span></a>
<a class="sourceLine" id="cb4-32" title="32"><span class="co"># 2   NA19139 YRI -55.41972 -19.32379</span></a>
<a class="sourceLine" id="cb4-33" title="33"><span class="co"># 3   NA18912 YRI -49.86350 -22.98661</span></a>
<a class="sourceLine" id="cb4-34" title="34"><span class="co"># 4   NA19160 YRI -58.62539 -22.75500</span></a>
<a class="sourceLine" id="cb4-35" title="35"><span class="co"># 5   NA07034 CEU  79.90126  73.59765</span></a>
<a class="sourceLine" id="cb4-36" title="36"><span class="co"># 6   NA07055 CEU  82.78208  75.75534</span></a></code></pre></div>
</div>
<div id="visualize-the-population-structure" class="section level3">
<h3 class="hasAnchor">
<a href="#visualize-the-population-structure" class="anchor"></a>Visualize the population structure</h3>
<p>Project the genetic features onto a plane.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="va">pop_str</span><span class="op">$</span><span class="va">RD1</span>, <span class="va">pop_str</span><span class="op">$</span><span class="va">RD2</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span><span class="op">(</span><span class="va">pop_str</span><span class="op">$</span><span class="va">pop</span><span class="op">)</span>, xlab<span class="op">=</span><span class="st">"RD 1"</span>, ylab<span class="op">=</span><span class="st">"RD 2"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">pop_str</span><span class="op">$</span><span class="va">pop</span><span class="op">)</span>, pch<span class="op">=</span><span class="st">"o"</span>, col<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nlevels.html">nlevels</a></span><span class="op">(</span><span class="va">pop_str</span><span class="op">$</span><span class="va">pop</span><span class="op">)</span><span class="op">)</span></pre></div>
<div class="figure" style="text-align: center">
<img src="Genome_scan_KLFDAPC_files/figure-html/figure%201-1.png" alt="&amp;nbsp;" width="768"><p class="caption">
 
</p>
</div>
</div>
<div id="doing-genome-scan-using-klfdapt" class="section level3">
<h3 class="hasAnchor">
<a href="#doing-genome-scan-using-klfdapt" class="anchor"></a>Doing genome scan using KLFDAPT</h3>
<p>After we choose the genetic features, we then use the genetic features to do genome scan with KLFDAPT.</p>
<p>?KLFDAPT to see the documentation. This genome scan is handeled with a C++ function, which is faster than other implementation. Users can also set the parallel process by changing the paramters of num.threshold.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"></a>
<a class="sourceLine" id="cb6-2" title="2"><span class="kw">showfile.gds</span>(<span class="dt">closeall=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb6-3" title="3"></a>
<a class="sourceLine" id="cb6-4" title="4"><span class="co"># Get chromosome index</span></a>
<a class="sourceLine" id="cb6-5" title="5"></a>
<a class="sourceLine" id="cb6-6" title="6">genofile &lt;-<span class="st"> </span><span class="kw">snpgdsOpen</span>(<span class="kw">snpgdsExampleFileName</span>())</a>
<a class="sourceLine" id="cb6-7" title="7"></a>
<a class="sourceLine" id="cb6-8" title="8">chr &lt;-<span class="st"> </span><span class="kw">read.gdsn</span>(<span class="kw">index.gdsn</span>(genofile, <span class="st">"snp.chromosome"</span>))</a>
<a class="sourceLine" id="cb6-9" title="9"></a>
<a class="sourceLine" id="cb6-10" title="10">klfdapcmat=<span class="kw">as.matrix</span>(pop_str[,<span class="op">-</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>)])</a>
<a class="sourceLine" id="cb6-11" title="11"></a>
<a class="sourceLine" id="cb6-12" title="12"><span class="co">### Doing genome scan</span></a>
<a class="sourceLine" id="cb6-13" title="13"><span class="co">### using the correlation to estimate the contribution of SNPs to the pop str (which is the same as PCA does )</span></a>
<a class="sourceLine" id="cb6-14" title="14"></a>
<a class="sourceLine" id="cb6-15" title="15">klfdapcscanqvalue &lt;-<span class="st"> </span><span class="kw">KLFDAPT</span>(klfdapcmat, genofile, <span class="dt">n.RD=</span><span class="dv">1</span><span class="op">:</span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb6-16" title="16"><span class="co"># Warning: package 'robust' was built under R version 3.6.3</span></a>
<a class="sourceLine" id="cb6-17" title="17"><span class="co"># Warning: package 'fit.models' was built under R version 3.6.3</span></a>
<a class="sourceLine" id="cb6-18" title="18"><span class="co"># SNP Correlation:</span></a>
<a class="sourceLine" id="cb6-19" title="19"><span class="co"># Working space: 279 samples, 9088 SNPs</span></a>
<a class="sourceLine" id="cb6-20" title="20"><span class="co">#     using 1 (CPU) core</span></a>
<a class="sourceLine" id="cb6-21" title="21"><span class="co">#     using the top 2 eigenvectors</span></a>
<a class="sourceLine" id="cb6-22" title="22"><span class="co"># Correlation:    the sum of all selected genotypes (0,1,2) = 2553065</span></a>
<a class="sourceLine" id="cb6-23" title="23"><span class="co"># Thu Oct 29 00:40:51 2020    (internal increment: 10216)</span></a>
<a class="sourceLine" id="cb6-24" title="24"><span class="co"># </span></a>
<a class="sourceLine" id="cb6-25" title="25">[..................................................]  <span class="dv">0</span>%, ETC<span class="op">:</span><span class="st"> </span><span class="op">---</span><span class="st">        </span></a>
<a class="sourceLine" id="cb6-26" title="26">[<span class="op">==</span><span class="er">================================================</span>] <span class="dv">100</span>%, completed, 0s</a>
<a class="sourceLine" id="cb6-27" title="27"><span class="co"># Thu Oct 29 00:40:51 2020    Done.</span></a></code></pre></div>
</div>
<div id="plot-manhattan-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-manhattan-plot" class="anchor"></a>Plot manhattan plot</h3>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="co">###plot manhattan plot </span>
 <span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">klfdapcscanqvalue</span><span class="op">$</span><span class="va">q.values</span><span class="op">$</span><span class="va">q.values</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">4</span><span class="op">)</span>, xlab<span class="op">=</span><span class="st">""</span>, ylab<span class="op">=</span><span class="st">"-log(q-value)"</span>,col<span class="op">=</span><span class="va">chr</span>, pch<span class="op">=</span><span class="st">"+"</span><span class="op">)</span>
   <span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="st">"blue"</span><span class="op">)</span></pre></div>
<div class="figure" style="text-align: center">
<img src="Genome_scan_KLFDAPC_files/figure-html/figure%202-1.png" alt="&amp;nbsp;" width="768"><p class="caption">
 
</p>
</div>
</div>
<div id="reference" class="section level3">
<h3 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference</h3>
<p>Patterson, N., Price, A. L., &amp; Reich, D. (2006). Population structure and eigenanalysis. PLoS genet, 2(12), e190.</p>
<p>Luu, K., Bazin, E., &amp; Blum, M. G. (2017). pcadapt: an R package to perform genome scans for selection based on principal component analysis. Molecular ecology resources, 17(1), 67-77.</p>
<p>A High-performance Computing Toolset for Relatedness and Principal Component Analysis of SNP Data. Xiuwen Zheng; David Levine; Jess Shen; Stephanie M. Gogarten; Cathy Laurie; Bruce S. Weir. Bioinformatics 2012; doi: 10.1093/bioinformatics/bts606.</p>
<p>R. A. Maronna and R. H. Zamar (2002) Robust estimates of location and dispersion of high-dimensional datasets. Technometrics 44 (4), 307–317.</p>
<p>Wang, J., Zamar, R., Marazzi, A., Yohai, V., Salibian-Barrera, M., Maronna, R., … &amp; Maechler, M. (2017). robust: Port of the S+“Robust Library”. R package version 0.4-18. Available online at: <a href="https://CRAN" class="uri">https://CRAN</a>. R-project. org/package= robust.</p>
<p>Qin, X., Chiang, C.W.K., and Gaggiotti, O.E. (2021). Kernel Local Fisher Discriminant Analysis of Principal Components (KLFDAPC) significantly improves the accuracy of predicting geographic origin of individuals. <a href="https://www.biorxiv.org/content/10.1101/2021.05.15.444294v2.full" class="uri">bioRxiv</a>, 2021.2005.2015.444294.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Xinghu Qin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
