<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Genome scan for detecting natural selection based on KLFDAPC
 — KLFDAPT • KLFDAPC</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Genome scan for detecting natural selection based on KLFDAPC
 — KLFDAPT" />
<meta property="og:description" content="KLFDAPT implements genome scan based on KLFDAPC. The implementation is similar with PCA-based genome scan that reported by Patterson, N.,et al. 2006 and Luu, K., et al. 2017. The loci are weighted by the correlations between genetic variation and the genetic structure (first K reduced features).
" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-158177382-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-158177382-1');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">KLFDAPC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../reference/index.html">Manual</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/xinghuq/KLFDAPC/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Genome scan for detecting natural selection based on KLFDAPC
<!-- %%  ~~function to do ... ~~ --></h1>
    
    <div class="hidden name"><code>KLFDAPT.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>KLFDAPT implements genome scan based on KLFDAPC. The implementation is similar with PCA-based genome scan that reported by Patterson, N.,et al. 2006 and Luu, K., et al. 2017. The loci are weighted by the correlations between genetic variation and the genetic structure (first K reduced features).
<!-- %%  ~~ A concise (1-5 lines) description of what the function does. ~~ --></p>
    </div>

    <pre class="usage"><span class='fu'>KLFDAPT</span><span class='op'>(</span><span class='va'>klfdapc</span>, <span class='va'>gdsfile</span>, estimation<span class='op'>=</span><span class='st'>"auto"</span>, snp.id <span class='op'>=</span> <span class='cn'>NULL</span>, n.RD <span class='op'>=</span> <span class='cn'>NULL</span>, num.thread <span class='op'>=</span> <span class='fl'>1L</span>, with.id <span class='op'>=</span> <span class='cn'>TRUE</span>, outgds <span class='op'>=</span> <span class='cn'>NULL</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>klfdapc</th>
      <td><p>KLFDAPC matrix. The results returned from KLFDAPC analysis.
<!-- %%     ~~Describe \code{klfdapc} here~~ --></p></td>
    </tr>
    <tr>
      <th>gdsfile</th>
      <td><p>The genotype file, it is an object of class SNPGDSFileClass, a SNP GDS file
<!-- %%     ~~Describe \code{gdsfile} here~~ --></p></td>
    </tr>
    <tr>
      <th>estimation</th>
      <td><p>a character string specifying the robust estimator to be used. The choices are: "mcd" for the Fast MCD algorithm of Rousseeuw and Van Driessen, "weighted" for the Reweighted MCD, "donostah" for the Donoho-Stahel projection based estimator, "M" for the constrained M estimator provided by Rocke, "pairwiseQC" for the orthogonalized quadrant correlation pairwise estimator, and "pairwiseGK" for the Orthogonalized Gnanadesikan-Kettenring pairwise estimator. The default "auto" selects from "donostah", "mcd", and "pairwiseQC" with the goal of producing a good estimate in a reasonable amount of time.
<!-- %%     ~~Describe \code{verbose} here~~ --></p></td>
    </tr>
    <tr>
      <th>snp.id</th>
      <td><p>a vector of snp id indicating the selected SNP id, if NULL, all SNPs are used
<!-- %%     ~~Describe \code{snp.id} here~~ --></p></td>
    </tr>
    <tr>
      <th>n.RD</th>
      <td><p>a vector of integers indicating which reduced features to be used in KLFDAPC
<!-- %%     ~~Describe \code{eig.which} here~~ --></p></td>
    </tr>
    <tr>
      <th>num.thread</th>
      <td><p>the number of (CPU) cores used; if NA, detect the number of cores automatically
<!-- %%     ~~Describe \code{num.thread} here~~ --></p></td>
    </tr>
    <tr>
      <th>with.id</th>
      <td><p>if TRUE, the returned value with sample.id and sample.id
<!-- %%     ~~Describe \code{with.id} here~~ --></p></td>
    </tr>
    <tr>
      <th>outgds</th>
      <td><p>NULL or a character of file name for exporting correlations to a GDS file, see details
<!-- %%     ~~Describe \code{outgds} here~~ --></p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>if TRUE, show information
<!-- %%     ~~Describe \code{verbose} here~~ --></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>For more details, please read our vignette (https://github.com/xinghuq/KLFDAPC).
<!-- %%  ~~ If necessary, more details than the description above ~~ --></p>
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><!-- %%  ~Describe the value returned -->
<!-- %%  If it is a LIST, use --></p>
<dt>cor</dt><dd><p>The correlation coefficients between the genotypes and the population genetic structure</p></dd>
 <dt>q.values </dt><dd><p>The p.values and q.values for the correlation coefficients</p></dd>
%% ...

    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Patterson, N., Price, A. L., &amp; Reich, D. (2006). Population structure and eigenanalysis. PLoS genet, 2(12), e190.</p>
<p>Luu, K., Bazin, E., &amp; Blum, M. G. (2017). pcadapt: an R package to perform genome scans for selection based on principal component analysis. Molecular ecology resources, 17(1), 67-77.</p>
<p>A High-performance Computing Toolset for Relatedness and Principal Component Analysis of SNP Data. Xiuwen Zheng; David Levine; Jess Shen; Stephanie M. Gogarten; Cathy Laurie; Bruce S. Weir. Bioinformatics 2012; doi: 10.1093/bioinformatics/bts606.</p>
<p>R. A. Maronna and R. H. Zamar (2002) Robust estimates of location and dispersion of high-dimensional datasets. Technometrics 44 (4), 307–317.</p>
<p>Wang, J., Zamar, R., Marazzi, A., Yohai, V., Salibian-Barrera, M., Maronna, R., ... &amp; Maechler, M. (2017). robust: Port of the S+“Robust Library”. R package version 0.4-18. Available online at: https://CRAN. R-project. org/package= robust.
<!-- %% ~put references to the literature/web site here ~ --></p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>qinxinghu@gmail.com
<!-- %%  ~~who you are~~ --></p>
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p><!-- %%  ~~further notes~~ --></p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><!-- %% ~~objects to See Also as \code{\link{help}}, ~~~ --></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>###Hap_map data</span>
<span class='co'>### Doing KLFDAPC, default kernel=kernlab::polydot(degree = 1, scale = 1, offset = 1)</span>
<span class='co'># Open the GDS file</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/xinghuq/KLFDAPC'>KLFDAPC</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://github.com/zhengxwen/SNPRelate'>SNPRelate</a></span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Loading required package: gdsfmt</span></div><div class='output co'>#&gt; <span class='message'>SNPRelate -- supported by Streaming SIMD Extensions 2 (SSE2)</span></div><div class='input'><span class='va'>genofile</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html'>snpgdsOpen</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/SNPRelate/man/snpgdsExampleFileName.html'>snpgdsExampleFileName</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>
<span class='co'># Get sample id</span>
<span class='va'>sample.id</span> <span class='op'>&lt;-</span> <span class='fu'>read.gdsn</span><span class='op'>(</span><span class='fu'>index.gdsn</span><span class='op'>(</span><span class='va'>genofile</span>, <span class='st'>"sample.id"</span><span class='op'>)</span><span class='op'>)</span>
<span class='co'># Get population information</span>

<span class='va'>pop_code</span> <span class='op'>&lt;-</span> <span class='fu'>read.gdsn</span><span class='op'>(</span><span class='fu'>index.gdsn</span><span class='op'>(</span><span class='va'>genofile</span>, <span class='st'>"sample.annot/pop.group"</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># assume the order of sample IDs is as the same as population codes</span>

<span class='va'>pop_code</span> <span class='op'>&lt;-</span> <span class='fu'>read.gdsn</span><span class='op'>(</span><span class='fu'>index.gdsn</span><span class='op'>(</span><span class='va'>genofile</span>, path<span class='op'>=</span><span class='st'>"sample.annot/pop.group"</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>pop_code</span><span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>pop_code</span>,levels<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>pop_code</span><span class='op'>)</span><span class='op'>)</span>
<span class='co'># fliter LD thresholds for sensitivity analysis</span>
<span class='va'>snpset</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SNPRelate/man/snpgdsLDpruning.html'>snpgdsLDpruning</a></span><span class='op'>(</span><span class='va'>genofile</span>, ld.threshold<span class='op'>=</span><span class='fl'>0.2</span><span class='op'>)</span>
</div><div class='output co'>#&gt; SNP pruning based on LD:
#&gt; Excluding 365 SNPs on non-autosomes
#&gt; Excluding 1 SNP (monomorphic: TRUE, MAF: NaN, missing rate: NaN)
#&gt; Working space: 279 samples, 8,722 SNPs
#&gt;     using 1 (CPU) core
#&gt;     sliding window: 500,000 basepairs, Inf SNPs
#&gt;     |LD| threshold: 0.2
#&gt;     method: composite
#&gt; Chromosome 1: 75.84%, 543/716
#&gt; Chromosome 2: 72.78%, 540/742
#&gt; Chromosome 3: 74.38%, 453/609
#&gt; Chromosome 4: 73.49%, 413/562
#&gt; Chromosome 5: 76.33%, 432/566
#&gt; Chromosome 6: 75.40%, 426/565
#&gt; Chromosome 7: 75.42%, 356/472
#&gt; Chromosome 8: 71.31%, 348/488
#&gt; Chromosome 9: 77.64%, 323/416
#&gt; Chromosome 10: 74.33%, 359/483
#&gt; Chromosome 11: 77.18%, 345/447
#&gt; Chromosome 12: 76.35%, 326/427
#&gt; Chromosome 13: 75.87%, 261/344
#&gt; Chromosome 14: 76.60%, 216/282
#&gt; Chromosome 15: 76.34%, 200/262
#&gt; Chromosome 16: 73.38%, 204/278
#&gt; Chromosome 17: 73.91%, 153/207
#&gt; Chromosome 18: 74.44%, 198/266
#&gt; Chromosome 19: 85.00%, 102/120
#&gt; Chromosome 20: 71.62%, 164/229
#&gt; Chromosome 21: 75.40%, 95/126
#&gt; Chromosome 22: 77.59%, 90/116
#&gt; 6,547 markers are selected in total.</div><div class='input'>
<span class='co'># Get all selected snp id</span>
<span class='va'>snpset.id</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/unlist.html'>unlist</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/unname.html'>unname</a></span><span class='op'>(</span><span class='va'>snpset</span><span class='op'>)</span><span class='op'>)</span>

<span class='fu'><a href='https://rdrr.io/pkg/SNPRelate/man/snpgdsClose.html'>snpgdsClose</a></span><span class='op'>(</span><span class='va'>genofile</span><span class='op'>)</span>

<span class='va'>klfdapcfile</span> <span class='op'>&lt;-</span> <span class='fu'><a href='KLFDAPC.html'>KLFDAPC</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/SNPRelate/man/snpgdsExampleFileName.html'>snpgdsExampleFileName</a></span><span class='op'>(</span><span class='op'>)</span>, y<span class='op'>=</span><span class='va'>pop_code</span>, snp.id<span class='op'>=</span><span class='va'>snpset.id</span>,n.pc <span class='op'>=</span> <span class='fl'>3</span>, kernel<span class='op'>=</span><span class='fu'>kernlab</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/kernlab/man/dots.html'>polydot</a></span><span class='op'>(</span>degree <span class='op'>=</span> <span class='fl'>1</span>, scale <span class='op'>=</span> <span class='fl'>1</span>, offset <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>, r<span class='op'>=</span><span class='fl'>5</span>,num.thread<span class='op'>=</span><span class='fl'>2</span><span class='op'>)</span>
</div><div class='output co'>#&gt; Principal Component Analysis (PCA) on genotypes:
#&gt; Excluding 2,541 SNPs (non-autosomes or non-selection)
#&gt; Excluding 0 SNP (monomorphic: TRUE, MAF: NaN, missing rate: NaN)
#&gt; Working space: 279 samples, 6,547 SNPs
#&gt;     using 2 (CPU) cores
#&gt; PCA:    the sum of all selected genotypes (0,1,2) = 1829478
#&gt; CPU capabilities: Double-Precision SSE2
#&gt; Thu Oct 29 00:40:43 2020    (internal increment: 1276)
#&gt; [..................................................]  0%, ETC: ---        [==================================================] 100%, completed, 0s
#&gt; Thu Oct 29 00:40:43 2020    Begin (eigenvalues and eigenvectors)
#&gt; Thu Oct 29 00:40:44 2020    Done.
#&gt; [1] "Doing KLFDAPC"</div><div class='output co'>#&gt; <span class='warning'>Warning: NaNs produced</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: WMDB</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: DA</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'DA'</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:KLFDAPC':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    KLFDA</span></div><div class='output co'>#&gt; <span class='warning'>Warning: no non-missing arguments to min; returning Inf</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: klaR</span></div><div class='output co'>#&gt; <span class='warning'>Warning: package 'klaR' was built under R version 3.6.3</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: MASS</span></div><div class='output co'>#&gt; <span class='warning'>Warning: collapsing to unique 'x' values</span></div><div class='output co'>#&gt; <span class='warning'>Warning: collapsing to unique 'x' values</span></div><div class='output co'>#&gt; <span class='warning'>Warning: collapsing to unique 'x' values</span></div><div class='output co'>#&gt; <span class='warning'>Warning: collapsing to unique 'x' values</span></div><div class='output co'>#&gt; <span class='error'>Error in density.default(xx, ...): 'x' contains missing values</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/pkg/SNPRelate/man/snpgdsClose.html'>snpgdsClose</a></span><span class='op'>(</span><span class='va'>genofile</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in closefn.gds(gdsobj): The GDS file is closed or uninitialized.</span></div><div class='input'><span class='co'># Make a data.frame</span>
<span class='va'>pop_str</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>sample.id <span class='op'>=</span> <span class='va'>klfdapcfile</span><span class='op'>$</span><span class='va'>PCA</span><span class='op'>$</span><span class='va'>sample.id</span>,
    pop <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>pop_code</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/match.html'>match</a></span><span class='op'>(</span><span class='va'>klfdapcfile</span><span class='op'>$</span><span class='va'>PCA</span><span class='op'>$</span><span class='va'>sample.id</span>, <span class='va'>sample.id</span><span class='op'>)</span><span class='op'>]</span>,
    RD1 <span class='op'>=</span> <span class='va'>klfdapcfile</span><span class='op'>$</span><span class='va'>KLFDAPC</span><span class='op'>$</span><span class='va'>Z</span><span class='op'>[</span>,<span class='fl'>1</span><span class='op'>]</span>,    <span class='co'># the first reduced feature</span>
    RD2 <span class='op'>=</span>  <span class='va'>klfdapcfile</span><span class='op'>$</span><span class='va'>KLFDAPC</span><span class='op'>$</span><span class='va'>Z</span><span class='op'>[</span>,<span class='fl'>2</span><span class='op'>]</span>,    <span class='co'># the second reduced feature</span>
    stringsAsFactors <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in data.frame(sample.id = klfdapcfile$PCA$sample.id, pop = factor(pop_code)[match(klfdapcfile$PCA$sample.id,     sample.id)], RD1 = klfdapcfile$KLFDAPC$Z[, 1], RD2 = klfdapcfile$KLFDAPC$Z[,     2], stringsAsFactors = FALSE): object 'klfdapcfile' not found</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>pop_str</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in head(pop_str): object 'pop_str' not found</span></div><div class='input'>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>pop_str</span><span class='op'>)</span><span class='op'>=</span><span class='va'>pop_str</span><span class='op'>$</span><span class='va'>sample.id</span>
</div><div class='output co'>#&gt; <span class='error'>Error in eval(expr, envir, enclos): object 'pop_str' not found</span></div><div class='input'>
<span class='va'>genofile</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html'>snpgdsOpen</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/SNPRelate/man/snpgdsExampleFileName.html'>snpgdsExampleFileName</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>chr</span> <span class='op'>&lt;-</span> <span class='fu'>read.gdsn</span><span class='op'>(</span><span class='fu'>index.gdsn</span><span class='op'>(</span><span class='va'>genofile</span>, <span class='st'>"snp.chromosome"</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>klfdapcmat</span><span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='va'>pop_str</span><span class='op'>[</span>,<span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>2</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in as.matrix(pop_str[, -c(1:2)]): object 'pop_str' not found</span></div><div class='input'>
<span class='co'>### using the correlation to estimate the contribution of SNPs to the pop str (which is the same as PCA does )</span>

<span class='va'>klfdapcscanqvalue</span> <span class='op'>&lt;-</span> <span class='fu'>KLFDAPT</span><span class='op'>(</span><span class='va'>klfdapcmat</span>, <span class='va'>genofile</span>, n.RD<span class='op'>=</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>2</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error: object 'klfdapcmat' not found</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/pkg/SNPRelate/man/snpgdsClose.html'>snpgdsClose</a></span><span class='op'>(</span><span class='va'>genofile</span><span class='op'>)</span>
<span class='co'>###Mmanhattan plot</span>

 <span class='fu'><a href='https://rdrr.io/r/graphics/plot.html'>plot</a></span><span class='op'>(</span><span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log10</a></span><span class='op'>(</span><span class='va'>klfdapcscanqvalue</span><span class='op'>$</span><span class='va'>q.values</span><span class='op'>$</span><span class='va'>q.values</span><span class='op'>)</span>, ylim<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,<span class='fl'>4</span><span class='op'>)</span>, xlab<span class='op'>=</span><span class='st'>""</span>, ylab<span class='op'>=</span><span class='st'>"-log(q-value)"</span>,col<span class='op'>=</span><span class='va'>chr</span>, pch<span class='op'>=</span><span class='st'>"+"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in plot(-log10(klfdapcscanqvalue$q.values$q.values), ylim = c(0,     4), xlab = "", ylab = "-log(q-value)", col = chr, pch = "+"): object 'klfdapcscanqvalue' not found</span></div><div class='input'>   <span class='fu'><a href='https://rdrr.io/r/graphics/abline.html'>abline</a></span><span class='op'>(</span>h<span class='op'>=</span><span class='fl'>2</span>, col<span class='op'>=</span><span class='st'>"blue"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in int_abline(a = a, b = b, h = h, v = v, untf = untf, ...): plot.new has not been called yet</span></div><div class='input'>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Xinghu Qin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


